server/app/api/v1/databases.py

```diff
@@ -45,13 +46,24 @@ async def connect_to_database(
     try:
         connection_string = f"postgresql://{database_user}:{database_password}@{database_host}:{database_port}/{database_name}"
 
-        user_database_obj = UserDatabase(name=f"{database_name}", user_id=current_user.id, connection_string=connection_string)
+        engine = create_engine(connection_string)
+        try:
+            engine.connect()
+        except Exception as e:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="Unable to connect to the database: " + str(e),
+            )
+        
+        fernet_manager = FernetManager(current_user.hashed_password)
```

Comment:
The key should derived using the user's plaintext password, not the hashed password. In order to derive the key from user's password, we can either do it at the login step and save the key as an HTTP cookie for subsequent request or, we can prompt the user to enter their password again when they click on the "Ask Query" button and then derive the key and decrypt the connection string. We can discuss this in detail.

Reply:
1. Approach one of storing in HTTP cookie is highly insecure and never recommended
2. Approach two is acceptable but adds friction during usage and might not be the best option

Instead we could generate a key the fernet manager creates and store it in db but that would be the same as using hashed password.  Ideally the keys should be stored separately in a vault but we do want to complicate the usage of this app for the user

Reply:
> * Approach one of storing in HTTP cookie is highly insecure and never recommended

Agree, and that is what I had mentioned in the [doc](https://gist.github.com/harshshurutech/78c8aa2b44d6acdd02ba8b824b2d41fa?permalink_comment_id=5033204#gistcomment-5033204).

> 2. Approach two is acceptable but adds friction during usage and might not be the best option

Yes, it does introduce some friction.

Another approach is to run a local vault service in the Docker container on the same network, which we can use for storing the keys. I am currently looking into it.

Reply:
Or docker works too for now. Let me know what you find

Reply:
[Infisical](https://infisical.com/docs/self-hosting/deployment-options/docker-compose) looks like a great open-source solution for secret management. I'll dig into it and keep you posted.

Reply:
Another option is [OpenBao](https://openbao.org/)